### DISCUSSION TITLE
[FR] Automatic search domain

### DISCUSSION THREAD

**Discussion by Anuskuss** [pivpn/pivpn#1387] [Feature requests]
**Is your feature request related to a problem? Please describe.**
When setting up a VPN server, pivpn should automatically detect the search domain and apply it.

**Describe the solution you'd like**
- If you're using Pi-hole, you can specify _Conditional forwarding_ which you can read from `/etc/dnsmasq.d/01-pihole.conf` (`server=/local/192.168.0.1`) or `/etc/pihole/setupVars.conf` (`REV_SERVER_DOMAIN=local`)
- If you're using the same DNS server as the host, localhost or whatever you can read it from `/etc/resolv.conf` (`search local`)

I'm not completely sure what the canonical way to specify the search domain in [WireGuard](https://github.com/pivpn/pivpn/blob/32bd1c628af9e1926f3f4471c0bf49c74deff7c7/scripts/wireguard/makeCONF.sh#L100) is, but this works:
```ini
[Interface]
DNS = 10.6.0.1, local
```

**Describe alternatives you've considered**
\-

**Additional context**
I've only tested that config on Windows 10, don't know if all clients support this.

**Comment by orazioedoardo** [comment]
Are you sure you can use `.local` over VPN? AFAIK if you try to connect to `mypc.local` it should reach the PC with the hostname mypc on the local LAN, not the remote one.

**Comment by Anuskuss** [comment]
I've used my phone for the connection (tethering) and was able to ping my `landevice.local` (or just `landevice` with my modification) and there's no way that that device exists inside my mobile data network.

**Comment by orazioedoardo** [comment]
> there's no way that that device exists inside my mobile data network.

Definitely. But what if the user connects to a non cellular network (for example someone else's wifi)? You could find some devices.

**Comment by Anuskuss** [comment]
But you would only find those if you're connected to the DNS server of "someone else's wifi". And pivpn disallowes local network in its default config anyway (`AllowedIPs = 0.0.0.0/0`) plus WireGuard has the *Block untunneled traffic* setting.

**Comment by orazioedoardo** [comment]
> But you would only find those if you're connected to the DNS server of "someone else's wifi".

[.local](https://en.wikipedia.org/wiki/.local) uses multicast DNS though, it doesn't go through a DNS server.

> And pivpn disallowes local network in its default config anyway (`AllowedIPs = 0.0.0.0/0`) plus WireGuard has the _Block untunneled traffic_ setting.

True, and for this reason if `mypc` exists in the local LAN, `mypc.local` will resolve to the local IP, and since we are tunnelling all traffic through VPN, it will connect to the same IP in the remote LAN, assuming it exists. So you can't connect to the remote `mypc` unless the VPN DNS takes precedence over mDNS (I'm not sure which is it).

By the way, I steered the conversation over .local, however other .something search domains should work.

**Comment by Anuskuss** [comment]
> .local uses multicast DNS though, it doesn't go through a DNS server.

I don't know, maybe you should only do this when the user chooses to use their Pi-hole. That's exactly how I've set up my own network, so it works for me.
If you agree with this, and also agree about the how, I'd say it should be implemented.

**Comment by isaacdcm** [comment]
New pivpn user here, but hitting the same issue.

I realize.local is a special domain name, but my dhcp gives every host a ".lan" which could have just as easily been ".thishomenetwork"  The inability to resolve names when I am vpn'ed in without explicitly adding that is kinda annoying.  Ive spent all morning looking for  a fix before landing here.

I assume (not an expert by any means) that the client goes through some sort of DHCP process to get an IP address, host name, and DNS settings.  There seems to be a custom domain option for openvpn which adds things to the domain search list, but I don't think this is available for wireguard, which I am using.
(I see commit here, cant find in documentation https://github.com/pivpn/pivpn/pull/706/commits/9808234792c0a3cef80bdafd6c5f00dc490d962b)

If the OP is on openvpn, that option may be the fix you are after, but still no solution for wireshark.

**Comment by Anuskuss** [comment]
> If the OP is on openvpn, that option may be the fix you are after, but still no solution for wireshark.

Well, I'm using WireGuard and there is a solution: I just change my configs like explained above (in your case `DNS = xxx.xxx.xxx.xxx, lan`). It's just that I want this done in pivpn so I (and others ofc) don't have to manually.

**Comment by Anuskuss** [comment]
FYI, the Android WireGuard client [recently](https://git.zx2c4.com/wireguard-android/commit/?id=3935a369b866c67705f3e27944be56b94ea2b245) gained support for search domains (using the `DNS = 1.2.3.4, lan` syntax).
It's easy enough to do that yourself, but I still think it would be beneficial for this to be read and applied automatically by PiVPN. But it may not be worth the trouble for you.

**Reply by coolapso** [reply]
Thanks,
PR's are welcome,
Discussions can be upvoted, Upvoting the feature helps us see which features are more requested :)

**Comment by clmcavaney** [comment]
It would be useful to have this added to pivpn.
Looking at the WireGuard makeCONF.sh at the moment - appears to be hard coded to only 2 DNS entries.
So, maybe the enhancement could be to change that to any number of DNS entries and then if a user specifies a domain search string (or strings) it could be joined as a comma separated list.

**Reply by coolapso** [reply]
Thanks, 
PR's are welcome,
Discussions can be upvoted, Upvoting the feature helps us see which features are more requested :)
